crypto/bio/bss_sock.c

```diff
@@ -151,6 +176,25 @@ static long sock_ctrl(BIO *b, int cmd, long num, void *ptr)
     case BIO_CTRL_FLUSH:
         ret = 1;
         break;
+    case BIO_CTRL_SET_KTLS:
+# ifndef OPENSSL_NO_KTLS
+        crypto_info = (struct tls12_crypto_info_aes_gcm_128 *)ptr;
```

Comment:
I've considered two possibilities to resolve this:
1. Move BIO_CTRL_SET_KTLS inside the "#ifndef OPENSSL_NO_KTLS" clause. Here and throughout. - **This is the approach I've implemented.**
2. Use a generic structure that will be converted to "struct tls12_crypto_info_aes_gcm_128" inside include/internal/ktls.h

Reply:
I don't see how that solves the problem. The issue is that you define `BIO_CTRL_SET_KTLS` and the related macro `BIO_set_ktls` in the public header file. As such you are declaring it as part of the public API and available for use by end user applications. However it is actually impossible for such applications to use because (as I understand it) `struct tls12_crypto_info_aes_gcm_128` is internal only so applications cannot make a sensible call to `BIO_set_ktls` anyway. In reality this seems to be an entirely internal thing, so these macros should probably not appear in the public headers at all. Probably this is true for some of the other macros too.

Reply:
I've relied on the code implemented in SCTP, where Linux specific code (internal structures) appear in dgram_ctrl.

I wouldn't mind to move these definitions out of the public header. But, it wouldn't be consistent to move BIO_CTRL_*KTLS to an internal header with other BIO_CTRL_* in the public header. AFAIU, there are other internal BIO_CTRL* that are already in the public header.

Reply:
Don't use the SCTP code as a model for how to do things!! OpenSSL has had a very long history and previously *everything* was in the public headers (there wasn't really the concept of internal headers). These days we are much stricter about such things. If it appears in the public headers then it is public API. If it isn't intended for public use then it should go into an internal header. I would put a comment in the public header saying that the BIO_CTRL numbers 72, 73, etc (or whatever the numbers are) are reserved for internal use, and then define them properly in include/internal/bio.h

Reply:
Thanks for the help. I'll move the definitions to include/internal/bio.h and leave a comment on the public file.